// @mysten/kiosk/dist/cjs/bcs.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bcs_exports = {};
__export(bcs_exports, {
  bcs: () => import_bcs.bcs
});
module.exports = __toCommonJS(bcs_exports);
var import_bcs = require("@mysten/sui.js/bcs");
var import_types = require("./types/index.js");
import_bcs.bcs.registerStructType(import_types.KIOSK_TYPE, {
  id: "address",
  profits: "u64",
  owner: "address",
  itemCount: "u32",
  allowExtensions: "bool"
});
import_bcs.bcs.registerStructType(import_types.KIOSK_PURCHASE_CAP, {
  id: "address",
  kioskId: "address",
  itemId: "address",
  minPrice: "u64"
});
import_bcs.bcs.registerStructType(import_types.TRANSFER_POLICY_CREATED_EVENT, {
  id: "address"
});
import_bcs.bcs.registerStructType(import_types.TRANSFER_POLICY_TYPE, {
  id: "address",
  balance: "u64",
  rules: ["vector", "string"]
});
